### PR TITLE
New Typed Signature Optimizer

### PR DESCRIPTION
**Pull request by thomasahle**
Also fixed a bug with list inputs, and another bug on predictors with no input fields.

### CODE REVIEW

**Comment by thomasahle** [comment]
I decided to not subclass `Teleprompter` and just make it a function.
Mostly because I didn't have any class state, and I couldn't see a reason for which arguments would go into the class constructor and which would go into `compile`.
Let me know if this is a problem.

**Comment by thomasahle** [comment]
I called it "Typed Signature Optimizer" because it assumes TypedPredictors rather than Predicts. Unfortunately the other signature optimizers don't work when using TypedPredictors, as the signatures get manipulated too much.

File: examples/nli/scone/ScoNe
Comment (okhat) [inline examples/nli/scone/ScoNe:1]: oh no a submodule haha. I hope it's not too hard to remove

Comment (thomasahle) [reply examples/nli/scone/ScoNe:1]: Oops, thanks for noticing!

**Pull request merged**